Added a way to list the product drive participants for a product drive

patelkrunal31 opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #3668 <!--fill issue number-->

### Description
- Updated logic to include participants as part of product_drive
- Updated participant model to count items donated for specific product_drive
<!-- Please include a summary of the change and which issue is fixed. 
Please also include relevant motivation and context.
Guide questions:
  - What motivated this change (if not already described in an issue)?
  - What alternative solutions did you consider?
  - What are the tradeoffs for your solution?
   
List any dependencies that are required for this change. (gems, js libraries, etc.)

Include anything else we should know about. -->

### Type of change

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?
- Added relevant tests to make sure correct number of items
<!-- Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Do we need to do anything else to verify your changes? 
If so, provide instructions (including any relevant configuration) so that we can reproduce? -->

### Screenshots
<img width="1528" alt="Screenshot 2023-07-21 at 5 53 44 PM" src="https://github.com/rubyforgood/human-essentials/assets/92335682/b9f5cd11-c000-41b6-94ff-c78482f25e18">

<!--Optional. Delete if not relevant. 
Include screenshots (before / after) for style changes, highlight edited element.-->

scooter-dangle commented on app/views/product_drives/show.html.erb:
```diff
@@ -100,3 +100,48 @@
+                <th>Email</th>
+                <th>Address</th>
+                <th>Items for this Drive</th>
+                <th>Total Items</th>
```
@cielf, do you think we could make it more clear that this is for all-time? (Maybe just something like
```suggestion
                <th>Total Items (for all drives)</th>
```

cielf replied:
Yeah, let's go with that -- we might refine it later.

scooter-dangle commented on app/models/product_drive_participant.rb:
```diff
@@ -31,6 +31,10 @@ def volume
   end
 
+  def volume_by_product_drive(product_drive_id)
+    donations.by_product_drive(product_drive_id).map { |d| d.line_items.total }.reduce(:+)
```
Minor: I think in general we'd want to default to `0` if we _somehow_ ended up with an empty array. Is that right? If so, would make sense to use `sum` rather than `reduce` to get that default value simply.

```suggestion
    donations.by_product_drive(product_drive_id).map { |d| d.line_items.total }.sum
```

scooter-dangle commented on spec/models/product_drive_participant_spec.rb:
```diff
@@ -35,14 +35,27 @@
 
+    describe "volume_by_product_drive" do
+      it "retrieves the amount of product that has been donated through specific product drive" do
+        dd1 = create(:product_drive)
+        dd2 = create(:product_drive)
+        ddp = create(:product_drive_participant)
+        create(:donation, :with_items, item_quantity: 10, source: Donation::SOURCES[:product_drive], product_drive: dd1, product_drive_participant: ddp)
+        create(:donation, :with_items, item_quantity: 9, source: Donation::SOURCES[:product_drive], product_drive: dd2, product_drive_participant: ddp)
+        expect(ddp.volume).to eq(19)
+        expect(ddp.volume_by_product_drive(dd1.id)).to eq(10)
+        expect(ddp.volume_by_product_drive(dd2.id)).to eq(9)
```
Nit: totes makes sense that you're copying the pre-existing variable naming scheme, but it's not a great naming scheme. Maybe `drive1`, `drive2`, and `participant` instead of `dd1`, `dd2`, and `ddp`?


```suggestion
        drive1 = create(:product_drive)
        drive2 = create(:product_drive)
        participant = create(:product_drive_participant)
        create(:donation, :with_items, item_quantity: 10, source: Donation::SOURCES[:product_drive], product_drive: drive1, product_drive_participant: participant)
        create(:donation, :with_items, item_quantity: 9, source: Donation::SOURCES[:product_drive], product_drive: drive2, product_drive_participant: participant)
        expect(participant.volume).to eq(19)
        expect(participant.volume_by_product_drive(drive1.id)).to eq(10)
        expect(participant.volume_by_product_drive(drive2.id)).to eq(9)
```

scooter-dangle reviewed (commented):
Left a couple comments, aside from which this looks good. Being a non Rails expert myself, maybe @awwaiid could give a quick skim for the Rails-idiomaticity?

patelkrunal31 commented:
> Left a couple comments, aside from which this looks good. Being a non Rails expert myself, maybe @awwaiid could give a quick skim for the Rails-idiomaticity?

I have addressed your feedback and it is ready for rereview.

cielf reviewed (commented):
LGTM as far as trying out the functionality goes.  Fine with it going forward if it passes technical review.  @awwaiid ?

dorner commented on app/models/product_drive_participant.rb:
```diff
@@ -31,6 +31,10 @@ def volume
   end
 
+  def volume_by_product_drive(product_drive_id)
+    donations.by_product_drive(product_drive_id).map { |d| d.line_items.total }.sum
```
There is almost definitely a more performant way to do this via SQL. Not sure if there tend to be more than a handful of product drive participants to be shown. @cielf ?

awwaiid replied:
Let's see how this goes, but I think this should be fine at the current scale. We would see slow load times on the page otherwise and can address them then.

awwaiid replied:
We have a total of ~17000 donations for all time

cielf replied:
And about 1/10 the number of product drive participants.

dorner commented on app/views/product_drives/show.html.erb:
```diff
@@ -100,3 +100,48 @@
   </div>
 </section>
+
+<section class="content">
```
Can we add a check in the request spec for this to see that the right information is being displayed? If there isn't a request spec, we should create one :)

patelkrunal31 replied:
I have added a test in my latest commit.

dorner commented:
Thanks for the contribution!
